### PR TITLE
feat(protocol): enable sp1 batch aggregation

### PR DESCRIPTION
**Pull request by smtmfft**

### CODE REVIEW

File: packages/protocol/contracts/layer1/verifiers/SP1Verifier.sol [hunk 1]
```diff
@@ -78,13 +79,54 @@ contract SP1Verifier is EssentialContract, IVerifier {
...
         external
-        pure
-        notImplemented
-    { }
+        view
+    {
+        // Extract the necessary data
+        bytes32 aggregation_program = bytes32(_proof.data[0:32]);
+        bytes32 block_proving_program = bytes32(_proof.data[32:64]);
+        bytes memory proof = _proof.data[64:];
```
Comment (dantaik) [inline packages/protocol/contracts/layer1/verifiers/SP1Verifier.sol:91]: if you use `_proof.data[64:]` directly on line122, maybe you can avoid mem-copy?

File: packages/protocol/contracts/layer1/verifiers/SP1Verifier.sol [hunk 2]
```diff
@@ -78,13 +79,54 @@ contract SP1Verifier is EssentialContract, IVerifier {
...
+        if (!isProgramTrusted[block_proving_program]) {
+            revert SP1_INVALID_PROGRAM_VKEY();
+        }
+
+        // Collect public inputs
+        bytes32[] memory public_inputs = new bytes32[](_ctxs.length + 1);
+        // First public input is the block proving program key
+        public_inputs[0] = block_proving_program;
+        // All other inputs are the block program public inputs (a single 32 byte value)
+        for (uint256 i = 0; i < _ctxs.length; i++) {
```
Comment (dantaik) [inline packages/protocol/contracts/layer1/verifiers/SP1Verifier.sol:107]:
```suggestion
        for (uint256 i; i < _ctxs.length; ++i) {
```

File: packages/protocol/contracts/layer1/verifiers/SP1Verifier.sol [hunk 3]
```diff
@@ -78,13 +79,54 @@ contract SP1Verifier is EssentialContract, IVerifier {
...
-        TaikoData.TierProof calldata /*_proof*/
+        ContextV2[] calldata _ctxs,
+        TaikoData.TierProof calldata _proof
     )
         external
-        pure
-        notImplemented
-    { }
+        view
+    {
```
Comment (dantaik) [inline packages/protocol/contracts/layer1/verifiers/SP1Verifier.sol:87]: maybe we shall check `_proof.dagta.length > 64`?

Comment (dantaik) [reply packages/protocol/contracts/layer1/verifiers/SP1Verifier.sol:87]: And `_ctxs.length !=0`?

**Review by Brechtpd — APPROVED** [review]

**Review by dantaik — APPROVED** [review]

**Review by YoGhurt111 — APPROVED** [review]

**Pull request merged**